crates/vdo/examples/basic.rs

```diff
@@ -0,0 +1,25 @@
+use anyhow::{Error, Result};
+use vdo::{Stream, VdoBufferStrategy, VdoFormat};
+
+fn main() -> Result<()> {
+    let mut stream = Stream::builder()
+        .channel(0)
+        .format(VdoFormat::VDO_FORMAT_PLANAR_RGB)
+        .width(1920)
+        .height(1080)
+        .buffer_strategy(VdoBufferStrategy::VDO_BUFFER_STRATEGY_EXPLICIT)
```

Comment:
The default VDO_BUFFER_STRATEGY_INFINITE is actually recommended. unfortunately we have not yet update all public examples, but this one uses the default infinite.

https://github.com/AxisCommunications/acap-native-sdk-examples/blob/main/vdostream/app/vdoencodeclient.c

instead of allocating buffers 1 by 1... you instead tell the server that we want "buffer.count" buffers, then there is no need for the allocate_buffers step, maybe it could be called "buffers"  when using infinite?

Reply:
Using VDO_BUFFER_STRATEGY_INFINITE is still the default if one were to call `Stream::new()`. I was just trying to demonstrate here how one might do things explicitly. Admittedly that is contrary to the "basic.rs" filename. I can adjust that.

I wonder though how "buffer.count" works in conjunction with `VDO_BUFFER_STRATEGY_INFINITE`. If we pass buffer.count = 5, and then don't enqueue any buffers until after we try to process a batch of 12, will that work or will it hang after it has provided the 5th frame without one being inquired again? Generally the purpose of "buffer.count" isn't clear to me.

Reply:
I see. yes, if you never return any buffers, it will hang if you exceed "buffer.count" frames. it's just a simpler alternative to explicit. since these buffers are physically contiguous they are a precious resource but only the application knows how many buffers are needed.

if you ask for 20 buffers the server might not be able to give you 20, the actual amount will be returned in the vdo_stream_get_info. so you can write a correct application that doesn't hang forever.

Reply:
I see. That makes more sense to me now. Thank you. I'm happy to make the adjustment.

Reply:
@daniel-nielsen-se I wonder about the internal reference counting of the VdoBuffer's a little bit. If the buffers are physically contiguous in memory and we use `VDO_BUFFER_STRATEGY_INFINITE` and call `vdo_stream_buffer_unref` on each buffer provided instead of `vdo_stream_buffer_enqueue`, how does the internal library handle allocations or indexing? 

Does it just allocate some default number of buffers, if we don't provide "buffer.count", and round-robin the queue?

My understanding right now is I ned to call `vdo_stream_get_buffer` followed by `vdo_stream_buffer_enqueue`, and when we're all done, call `vdo_stream_buffer_unref` for each buffer. I noticed in the example you linked, `vdo_stream_buffer_enqueu` was never called. It seems that if I provide "buffer.count" then I'm required to call `vdo_stream_buffer_enqueue`, but if "buffer.count" is not provided, that's not the case.

I think I'm getting the picture, but is my understanding correct so far and is that the intended operation? I may want to rethink the Rust API here a bit since the functionality is pretty different depending on which buffer strategy is used.

Reply:
Good question. The C api is complex because we try to avoid breaking changes.

We are aiming to deprecate all strategies except the default, so unless you really need it, as this is a "new start", it's probably better to not provide a wrapper for the old strategies. (If you have an important use-case for EXPLICIT, please let us know, then we can work towards improving infinite for these use-cases). (It used to be the case that explicit was faster which is why it exists at all, but that is no longer the case as we optimized infinite to be just as fast only simpler to use). 

Infinite will automatically cache mmap so one can simply call vdo_buffer_get_data. (never vdo_frame_memmap/vdo_frame_unmap).

**1) For YUV/RGB the default "buffer.count" is 3.**
**2) For jpeg/h26x etc. "buffer.count" is ignored.**

**3) vdo_stream_buffer_enqueue is scheduled to be deprecated.**
(because it's too easy to write incorrect code)

The only correct sequence is
**vdo_stream_buffer_enqueue**(...buffer...)
// not allowed to use buffer object after enqueue in anyway way except...
**g_clear_object**(&buffer); // g_object_unref or g_autoptr are equivalent.

So the recommended way is to use
**vdo_stream_buffer_unref**
which internally is simply implemented as the above sequence
**vdo_stream_buffer_enqueue+g_clear_object**
as the buffer variable is set to NULL it can no longer be misused by apps.

But just as you point out, that doesn't work for legacy strategies, which is another reason we want to deprecate them. Also infinite works for all formats including h26x, explicit works only for yuv/rgb etc.

For optimization reasons, internally we keep one extra reference for every buffer in the buffer pool, so when you call unref the count is simply reduced from 2 to 1.

**Short summary**
vdo_stream_buffer_enqueue is required for EXPLICIT.
(for infinite it doesn't matter if you set 'buffer.count' or not, you can always use vdo_stream_buffer_unref)

Reply:
That's some fantastic background and context. Thank you!

I agree with your perspective on using this as a new start and standardizing around `VDO_BUFFER_STRATEGY_INFINITE`. I've made that update to the wrapper. I've also kept `vdo_stream_buffer_unref` in the Drop trait implementation for `StreamBuffer`, so that is automatically called when it goes out of scope (i.e. each iteration of a loop).

I've tested with a couple formats and everything seems to be working well.